aten/src/ATen/native/mps/operations/Quantized.mm

```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+        ? ((short)tiitg/THREAD_PER_ROW_A)
+        : n_rows_A - 1;
+    short thread_row_B = ((short)tiitg/THREAD_PER_ROW_B) < n_rows_B
+        ? ((short)tiitg/THREAD_PER_ROW_B)
+        : n_rows_B - 1;
+
+    Tsimd8x8 ma[2]; // input
+    simdgroup_bfloat8x8 mb[4]; // weight
+    simdgroup_float8x8 c_res[8]; // outer product result
+    for (short i = 0; i < 8; i++){
+        c_res[i] = make_filled_simdgroup_matrix<float, 8>(0.f);
+    }
+
+    constant T * a_ptr = (constant T *)((constant char *)A
+        + nb11 * (r1 * BLOCK_SIZE_M + thread_row_A)
+        + nb10 * (BLOCK_SIZE_K / THREAD_PER_ROW_A * (tiitg % THREAD_PER_ROW_A)));
+
+    constant W * b_ptr = (constant W *)(B
+        + nb01 * (r0 * BLOCK_SIZE_N + thread_row_B)
+        + nb00 * (BLOCK_SIZE_K / THREAD_PER_ROW_B * (tiitg % THREAD_PER_ROW_B)));
+/**
+Load weight and input into shared memory:
+
+             K
+  ┌────────────────────────┐              8192 (A)           4096 (B)
+  │                        │   ┌────────────────────────┬────────────┐
+  │                        │   │++++++++++++++++++++++++│++++++++++++│
+  │                        │   └────────────────────────┴────────────┘
+  │                        │
+  │32                      │
+  ├──┬──┬──────────────────┤                    K
+  │++│  │                  │        ┌────────────────────────┐
+64│++│  │  ...             │        │                        │
+  │++│  │                  │        │                        │
+  ├──┴──┴──────────────────┤        │                        │
+  │                        │        │                        │
+  │      ───────────►      │        │32                      │
+  │       for loop         │        ├──┬──┬──────────────────┤
+  │                        │      32│++│  │ ...              │
+  │                        │        ├──┴──┴──────────────────┤
+  │                        │        │         ─────────────► │
+  │                        │        │            for loop    │
+  └────────────────────────┘        └────────────────────────┘
+              B                                 A
+ */
+    for (uint32_t loop_k = 0; loop_k < K; loop_k += BLOCK_SIZE_K) {
+        // load data and store to threadgroup memory
+        threadgroup_barrier(mem_flags::mem_threadgroup);
+
+        #pragma unroll(16)
+        for (short i = 0; i < 16; i++) {
+            bfloat weight = *(b_ptr + i);
+            // for example, tiitg 32, i 12 -> 0 + 1 = 1, it needs to work on sg mat grid row 1
+            short sg_mat_grid_row_index = (tiitg % THREAD_PER_ROW_B) * THREAD_PER_ROW_B + i / 8;
+            // same example, sg mat grid col index: 32 / 2 / 8 = 2, so currently need to work with sg mat at (1, 2)
+            short sg_mat_grid_col_index = tiitg / THREAD_PER_ROW_B / 8;
+            // now inside sg mat, which index to write to? starting point is SG_MAT_SIZE * sg_mat_offset
+            short row_offset = i & 7;
```

Comment:
nit: `i % 8` really should get you the same result from any reasonable compiler

Reply:
To be frank, I prefer `n & mask` to `n % (mask+1)`, because I remember that failed to do this optimization